### ISSUE TITLE
tutorial: why and how to write your own NixOS module

### ISSUE THREAD

**Issue by a-kenji** [NixOS/nix.dev#337]
## Problem

It might not immediately be apparent to users why to write their own NixOS modules and how to do it.

## Checklist

- [x] checked [latest Nixpkgs manual] \([source][nixpkgs-source]) and [latest NixOS manual] \([source][nixos-source])
- [x] checked [open documentation issues] for possible duplicates
- [x] checked [open documentation pull requests] for possible solutions

[latest Nixpkgs manual]: https://nixos.org/manual/nixpkgs/unstable/
[latest NixOS manual]: https://nixos.org/manual/nixos/unstable/
[nixpkgs-source]: https://github.com/NixOS/nixpkgs/tree/master/doc
[nixos-source]: https://github.com/NixOS/nixpkgs/tree/master/nixos/doc/manual
[open documentation issues]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+label%3A%229.needs%3A+documentation%22
[open documentation pull requests]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%228.has%3A+documentation%22%2C%226.topic%3A+documentation%22

---
Co-authored-by: @mightyiam

**Comment by fricklerhandwerk** [comment]
Moved this to nix.dev, as this is where we would maintain such a piece of documentation.

**Comment by hugosenari** [comment]
My personal history with nix:
1. Configuration language is hard
2. [Let me try fix it myself](https://github.com/cruel-intentions/stylops#related)
3. Then I discovered Nix Language (without modules) - at that moment I don't care about pkgs/drv
4. Then I discovered Nix Modules :exploding_head:

As today, modules can be used in:
- home manager
- devenv/devshell
- flake-parts
- We could use it everywhere a config file is required :heart: 

It expands our 'why', and that's why I created [devshell-files](https://github.com/cruel-intentions/devshell-files#about) (more modules) there are some why in about section.

And in _writing your modules section_: "it hides complexity, [hiding complexity is what abstraction is all about](http://mourlot.free.fr/english/fmtaureau.html), we didn't share options definitions across projects to type less, but because we could reuse an abstraction that helps hiding complexity."

Maybe we could add some why after #645 got merged.

@DavHau could give us some words to fill this why section?
